Declarations inside conditionals

STRd6 opened the pull request:
#444

- [x] Hoist Refs
- [x] Expand to while loops

edemaine commented:
Oh sweet! I was aiming to get here as well, so let me mention what my plan was so we can avoid overlap:

I have an unpushed branch that's refactoring statements into their own AST node (instead of `[indent, content, delimitet]`). I was thinking this would be helpful for walking up the tree from a declaration in an expression to find the containing statement and put the declaration there. So I was focusing on that first, and hadn't worked on actual declaration moving yet.

Would Statement AST nodes be useful for this, or did you have another way? I can hopefully get the Statement thing working in another couple of days.

STRd6 commented:
Statement AST wouldn't be too useful for this specifically, I just ended up using the statement tuple style.

Also refactored some of the block prefix code so some for loops have better indentation now. I removed the empty assignment hack since now we are tracking the actual assignments inside for loop headers.

Last thing is to add ref-hoisting.

The statement AST updates should fit in fine with this, I'll just need to adjust a couple places where we're inserting the indent.

edemaine commented:
I guess my question was: how are you planning to do ref hoisting? I was thinking it would make sense to follow parent pointers until you hit a Statement. But maybe you can do the same thing with Blocks?

STRd6 commented:
I was going to find the `parent` of the `IfStatement` and then splice a the hoisted ref statement before the `IfStatement` in the parent's children.

We'll probably need to do a whole additional pass on expressionizing statements later and I think the Stamement AST nodes could help a lot with that.

edemaine commented on test/if.civet:
Could you add a test of an if expression with a declaration?

STRd6 replied:
I don't think it works in expressions but I'll add a test with TODO indicating that.

edemaine commented on test/for.civet:
```diff
@@ -121,7 +120,7 @@ describe "for", ->
     ---
     for (let i1 = 1; i1 <= 10; ++i1) {
-    var i = i1;
+      var i = i1;
```
Thanks for fixing these!

edemaine commented on source/lib.js:
```diff
@@ -6,6 +6,50 @@
+ */
+function blockWithPrefix(clause, block) {
+  if (!clause) return block
+  const { blockPrefix } = clause
```
This API feels a little weird, but I guess it leaves room for passing in content other than `blockPrefix`? It might be cleaner to just pass in `clause?.blockPrefix` though.

From the name "blockWithPrefix", I'd also expect the arguments to be in the other order.

edemaine commented on source/lib.js:
```diff
@@ -6,6 +6,50 @@
+      return node
+    }
+  } while (node = node.parent)
+}
```
This is a special case of [findAncestor](https://github.com/DanielXMoore/Civet/blob/9b28e3a351d130ce031c7b28b50bcb6d2a499315/source/parser.hera#L7376). I'm fine with a separate implementation that's more efficient, but perhaps we could name it similarly (`findAncestorOfType` might be more descriptive), and maybe move `findAncestor`'s code to be next to this in lib?

STRd6 replied:
I named this one after the https://developer.mozilla.org/en-US/docs/Web/API/Element/closest API and eventually want to add full query selector support. `findAncestorOfType` is a better description of what it is doing currently, however.

edemaine commented on source/lib.js:
```diff
@@ -129,6 +183,26 @@ function hasYield(exp) {
+    .forEach(node => {
+      const { hoistableDec } = node
+
+      // TODO: expand set to include other parents that can have hoistable decs attached
```
Yeah, I worry that currently a declaration in an expression that is deeply inside an `if` body (with a function declaration or various other things in between) will end up attaching to the `if` here. But reading below, I guess that won't happen with the current grammar.  In the future I guess this is where `Statement` AST nodes would be helpful, and would enable declarations in more general places in expressions.

STRd6 replied:
👍 I imagine this will get refined a bit as we fill in more refs and do more with blocks and hoisting.

edemaine commented on source/parser.hera:
```diff
@@ -2921,22 +2933,31 @@ IfStatement
       then: block,
       else: e,
       children,
+      condition: clause.condition,
```
Would it make sense to instead start this object literal with `...clause`? Then could also omit `type` here.

edemaine commented on source/parser.hera:
```diff
@@ -3628,7 +3640,22 @@ CatchParameter
+      children: [open, expression, close],
+      expression,
+    }
+
```
Hmm, I expected this to be allowed in an arbitrary Expression, so you could do things like `if x + (y := ...)` but I guess that can be an extension for later.

edemaine commented on source/parser.hera:
```diff
@@ -3639,6 +3666,24 @@ Condition
+    const initCondition = {
+      type: "AssignmentExpression",
+      children: [ref, " ", initializer],
+      hoistableDec: [["", ["let ", ref], ";\n"]],
```
"Hoistable" suggests that it *could* be hosted but doesn't need to be. Maybe `hoistDec` would make more sense?

edemaine reviewed (approved):
Looks great! I just have a few minor comments.
